File: algorithm/eratosthenes.cc
// https://atcoder.jp/contests/abc084/tasks/abc084_d

#include <iostream>
#include <vector>
using namespace std;

// int(sqrt(N))  O(log(N))
int int_sqrt(int N) {
  int left = 0;
  int right = N;
  while (right - left > 1) {
    int middle = (left + right) / 2;
    if (middle * middle >= N) {
      right = middle;
    } else {
      left = middle;
    }
  }
  if (right * right >= N) {
    return right;
  } else {
    return left;
  }
}

// Sieve of Eratosthenes  O(N * sqrt(N))
vector<bool> eratosthenes(int N) {
  vector<bool> is_prime(N + 1, true);
  is_prime.at(0) = false;
  if (N > 0) {
    is_prime.at(1) = false;
  }
  for (int k = 2; k < int_sqrt(N); k++) {
    for (int i = 2 * k; i <= N; i+=k) {
      is_prime.at(i) = false;
    }
  }
  return is_prime;
}

int main() {
  vector<bool> is_prime = eratosthenes(100001);
  vector<bool> is_2017_like(100001, false);
  for (int n = 1; n < 100001; n+=2) {
    if (is_prime.at(n) && is_prime.at((n + 1) / 2)) {
      is_2017_like.at(n) = true;
    }
  }
  vector<int> sum(100001);
  sum.at(0) = 0;
  for (int n = 1; n < 100001; n++) {
    if (is_2017_like.at(n)) {
      sum.at(n) = sum.at(n - 1) + 1;
    } else {
      sum.at(n) = sum.at(n - 1);
    }
  }
  int Q;
  cin >> Q;
  for (int q = 0; q < Q; q++) {
    int l, r;
    cin >> l >> r;
    cout << (sum.at(r) - sum.at(l - 1)) << endl;
  }
  return 0;
}

File: algorithm/eratosthenes.py
INF = 9**18+4;  # Please set INF >= max(n) + 1.
def int_sqrt(n):
    if n == 0:
        return 0
    left = 0
    right = INF
    while right - left > 1:
        mid = (left + right) // 2;
        y = mid * mid - n;
        if y <= 0:
            left = mid
        else:
            right = mid
    return left

# Sieve of Eratosthenes  O(N * log log (N))
M = 3*(10**6)
is_prime = [True] * (M + 1)
def eratosthenes():
  is_prime[0] = False
  is_prime[1] = False
  for k in range(2, int_sqrt(M) + 1):
      for i in range(2 * k, M + 1, k):
          is_prime[i] = False

def main():
    eratosthenes()
    primes = []
    for m in range(2, M + 1):
        if is_prime[m]:
            primes.append(m)

    T = int(input())
    for t in range(T):
        # testcase
        N = int(input())
        ok = False
        # Fix p
        for p in primes:
            if p * p > N:
                break
            q = N // p // p
            if p * p * q != N:
                continue
            print(p, q)
            ok = True
            break
        if ok:
            continue
        # Fix q
        for q in primes:
            p2 = N // q
            p = int_sqrt(p2)
            if p * p * q != N:
                continue
            print(p, q)
            ok = True
        if ok:
            continue
        assert(False)

if __name__ == '__main__':
    main()
